### PR TITLE
remove go routine in recursion and track recursion depth

### PR DESCRIPTION
**Pull request by Oluwawunmi**
### What

remove go routine in recursion and track recursion depth

### How to review

Check code is ok.

### Who can review

Anyone, not me

### CODE REVIEW

File: cache/private/private_topic_test.go [hunk 1]
```diff
@@ -101,12 +112,24 @@ var (
 		SubtopicIds: &[]string{"1834"},
 	}
 
+	testEconomyRootTopic2 = models.Topic{
```
Comment (redhug1) [inline cache/private/private_topic_test.go:115]: Is this section duplicated in the public_test ?
and the next block of changes ?

**Review by redhug1 — COMMENTED** [review]
some code not copied ?

File: cache/private/private_topic_test.go (hunk 1 above)
Comment (Oluwawunmi) [reply cache/private/private_topic_test.go:115]: we've spoken about this.

File: cache/private/private_topic.go [hunk 2]
```diff
@@ -98,29 +95,39 @@ func UpdateDataTopics(ctx context.Context, serviceAuthToken string, topicClient
 	}
 }
 
-func processTopic(ctx context.Context, serviceAuthToken string, topicClient topicCli.Clienter, topicID string, topics *[]*cache.Topic, processedTopics map[string]bool, parentTopicID string) {
-	// Check if the topic is already processed
-	if processedTopics[topicID] {
+func processTopic(ctx context.Context, serviceAuthToken string, topicClient topicCli.Clienter, topicID string, topics *[]*cache.Topic, processedTopics map[string]struct{}, parentTopicID string, depth int) {
+	log.Info(ctx, "Processing topic at depth", log.Data{"topic_id": topicID,
+		"Depth": depth})
```
Comment (gedge) [inline cache/private/private_topic.go:100]:
```suggestion
	log.Info(ctx, "Processing topic at depth", log.Data{
	        "topic_id": topicID,
		"depth": depth,
	})
```

File: cache/private/private_topic.go [hunk 3]
```diff
@@ -98,29 +95,39 @@ func UpdateDataTopics(ctx context.Context, serviceAuthToken string, topicClient
...
+
+	// Check if the topic has already been processed
+	if _, exists := processedTopics[topicID]; exists {
+		err := errors.New("topic already processed")
+		log.Error(ctx, "Skipping already processed topic", err, log.Data{
+			"topic_id": topicID,
```
Comment (gedge) [inline cache/private/private_topic.go:106]:
```suggestion
		log.Error(ctx, "Skipping already processed topic", err, log.Data{
			"topic_id": topicID,
			"depth": depth,
```

File: cache/private/private_topic.go [hunk 4]
```diff
@@ -98,29 +95,39 @@ func UpdateDataTopics(ctx context.Context, serviceAuthToken string, topicClient
...
 		return
 	}
 
 	// Get the topic details from the topic client
 	dataTopic, err := topicClient.GetTopicPrivate(ctx, topicCli.Headers{ServiceAuthToken: serviceAuthToken}, topicID)
 	if err != nil {
-		log.Error(ctx, "failed to get topic details from topic-api", err)
+		log.Error(ctx, "failed to get topic details from topic-api", err, log.Data{
+			"Topic ID": topicID,
+			"Depth":    depth,
```
Comment (gedge) [inline cache/private/private_topic.go:116]:
```suggestion
			"topic_id": topicID,
			"depth":    depth,
```

File: cache/private/private_topic.go [hunk 5]
```diff
@@ -155,79 +162,66 @@ func getRootTopicCachePrivate(ctx context.Context, serviceAuthToken string, subt
...
 	rootTopicCache.Query = subtopicsIDMap.GetSubtopicsIDsQuery()
 
 	return rootTopicCache
 }
 
-func getSubtopicsIDsPrivate(ctx context.Context, serviceAuthToken string, subtopicsChan chan models.TopicResponse, topicClient topicCli.Clienter, topLevelTopicID string) {
-	topicCliReqHeaders := topicCli.Headers{ServiceAuthToken: serviceAuthToken}
+func processSubtopicsPrivate(ctx context.Context, serviceAuthToken string, subtopicsIDMap *cache.Subtopics, topicClient topicCli.Clienter, topLevelTopicID string, processedTopics map[string]struct{}, depth int) {
+	log.Info(ctx, "Processing topic at depth", log.Data{"topic_id": topLevelTopicID,
+		"Depth": depth})
```
Comment (gedge) [inline cache/private/private_topic.go:177]: as above

File: cache/private/private_topic.go [hunk 6]
```diff
@@ -155,79 +162,66 @@ func getRootTopicCachePrivate(ctx context.Context, serviceAuthToken string, subt
...
 
-	// get subtopics from dp-topic-api
-	subTopics, err := topicClient.GetSubtopicsPrivate(ctx, topicCliReqHeaders, topLevelTopicID)
+	// Check if this topic has already been processed
+	if _, exists := processedTopics[topLevelTopicID]; exists {
+		err := errors.New("topic already processed")
+		log.Error(ctx, "Skipping already processed topic", err, log.Data{
+			"topic_id": topLevelTopicID,
```
Comment (gedge) [inline cache/private/private_topic.go:183]: as above, add `depth`

File: cache/private/private_topic.go [hunk 7]
```diff
@@ -155,79 +162,66 @@ func getRootTopicCachePrivate(ctx context.Context, serviceAuthToken string, subt
...
 		return
 	}
 
-	// dereference sub-topics items to allow ranging through them
+	// Dereference sub-topics items to allow ranging through them
+	var subTopicItems []models.TopicResponse
 	if subTopics.PrivateItems == nil {
 		err := errors.New("items is nil")
-		log.Error(ctx, "failed to dereference sub-topics private items pointer", err, log.Data{"topic_id": topLevelTopicID})
+		log.Error(ctx, "failed to dereference sub-topics items pointer", err)
```
Comment (gedge) [inline cache/private/private_topic.go:208]: why remove this?

File: cache/public/public_topic.go [hunk 8]
```diff
@@ -96,9 +92,16 @@ func UpdateDataTopics(ctx context.Context, topicClient topicCli.Clienter) func()
 	}
 }
 
-func processTopic(ctx context.Context, topicClient topicCli.Clienter, topicID string, topics *[]*cache.Topic, processedTopics map[string]bool, parentTopicID string) {
-	// Check if the topic is already processed
-	if processedTopics[topicID] {
+func processTopic(ctx context.Context, topicClient topicCli.Clienter, topicID string, topics *[]*cache.Topic, processedTopics map[string]struct{}, parentTopicID string, depth int) {
+	log.Info(ctx, "Processing topic at depth", log.Data{"topic_id": topicID,
+		"Depth": depth})
```
Comment (gedge) [inline cache/public/public_topic.go:97]: as above (indent, lowercase)

File: cache/public/public_topic.go [hunk 9]
```diff
@@ -96,9 +92,16 @@ func UpdateDataTopics(ctx context.Context, topicClient topicCli.Clienter) func()
...
+
+	// Check if the topic has already been processed
+	if _, exists := processedTopics[topicID]; exists {
+		err := errors.New("topic already processed")
+		log.Error(ctx, "Skipping already processed topic", err, log.Data{
+			"topic_id": topicID,
```
Comment (gedge) [inline cache/public/public_topic.go:103]: as above - add `depth`

File: cache/public/public_topic.go [hunk 10]
```diff
@@ -107,6 +110,7 @@ func processTopic(ctx context.Context, topicClient topicCli.Clienter, topicID st
 	if err != nil {
 		log.Error(ctx, "failed to get topic details from topic-api", err, log.Data{
 			"Topic ID": topicID,
+			"Depth":    depth,
```
Comment (gedge) [inline cache/public/public_topic.go:113]: as above, make consistent keys

File: cache/public/public_topic.go [hunk 11]
```diff
@@ -156,55 +160,46 @@ func getRootTopicCachePublic(ctx context.Context, subtopicsChan chan models.Topi
...
-	// get subtopics from dp-topic-api
+func processSubtopicsPublic(ctx context.Context, subtopicsIDMap *cache.Subtopics, topicClient topicCli.Clienter, topLevelTopicID string, processedTopics map[string]struct{}, depth int) {
+	log.Info(ctx, "Processing topic at depth", log.Data{"topic_id": topLevelTopicID,
+		"Depth": depth})
+
+	// Check if this topic has already been processed
+	if _, exists := processedTopics[topLevelTopicID]; exists {
+		err := errors.New("topic already processed")
+		log.Error(ctx, "Skipping already processed topic", err, log.Data{
+			"topic_id": topLevelTopicID,
```
Comment (gedge) [inline cache/public/public_topic.go:181]: add depth?

**Review by gedge — COMMENTED** [review]
mostly needs consistent keys for logging

**Review by gedge — APPROVED** [review]

**Pull request merged**